Example of Jaiqu using burr

skrawcz opened the pull request:
Hi I saw [your hn post](https://news.ycombinator.com/item?id=39881753) and thought it would be good candidate for burr!

See https://news.ycombinator.com/item?id=39917364

Feedback / thoughts welcome! Basically the mermaid diagram enticed me to try burr since the mapping should be pretty close -- I think I got to a reasonable spot -- but it could be better. 

For those coming to view -- compare this with what's in the `jaiqu.py` file.

Here's a run in the burr UI:

![Screen Shot 2024-04-03 at 10 45 53 PM](https://github.com/AgentOps-AI/Jaiqu/assets/2328071/2034fe26-b14e-4340-9c83-d7dfd134ad99)

The two versions I created:

1. Straight port of existing code
![jaiqu_port](https://github.com/AgentOps-AI/Jaiqu/assets/2328071/a0f9a175-1e1c-468b-b4db-80f951925ef3)

2. More granular port trying to get close to the mermaid diagram
![jaiqu_granular](https://github.com/AgentOps-AI/Jaiqu/assets/2328071/9c69cc04-2549-4cb3-8649-7847f84ad328)

skrawcz committed:
Translating to burr

Step (1) big actions.
Step (2) write smaller actions that mirror the flow chart.

skrawcz committed:
Adds two implementations of Jaiqu using Burr

jaiqu_burr_straight_port takes the code and
maps it directly to burr.

jaiqu_burr_granular takes the straight
port and breaks it up into more actions.

Note: I was hoping to use the mermaid
diagram and do something reflecting that
but the logic in the mermaid diagram
doesn't really map to the code and
thus isn't a good thing to model -- but
the more granualr version does kind of
look like it.

skrawcz commented on jaiqu/__init__.py:
```diff
@@ -1 +1 @@
-from .jaiqu import validate_schema, translate_schema
+# from .jaiqu import validate_schema, translate_schema
```
me messing around...

areibman replied:
Keep or delete?

areibman commented:
@skrawcz seems you've made a lot of changes. Can you clarify/amend:

1. Why are there so many .xml files? Are they necessary for this?
2. Could you either add a lint rc into the repo or undo all the changes where whitespace is getting changed but not affecting the code?
3. Update gitignore for files that don't need to be in there
4. I'd prefer a direct but optional implementation on top of Jaiqu.py if you're going to add Burr. It doesn't seem like a good idea to copy + paste 1 file into 3 separate ones. That, or create an examples directory for each.
5. Which implementation is supposed to be better? Granular or straight? 
6. Would prefer any .PNGs just be rendered in README rather than just left in repo

Looks like a great idea, and it'd do a lot to help allay the complexity of Jaiqu.

areibman commented on .idea/.gitignore:
Accidentally added to git?

skrawcz replied:
yep!

areibman commented on jaiqu/helpers.py:
Undo whitespace

areibman commented on jaiqu/jaiqu:
Why is there no filename extension for this?

areibman commented on jaiqu/jaiqu.png:
Add to README in place of mermaid. (Mermaid was made by GPT).

areibman commented on jaiqu/jaiqu.py:
```diff
@@ -114,5 +118,5 @@ def translate_schema(input_json, output_schema, openai_api_key: str | None = Non
                     raise RuntimeError(f"Failed to validate the jq filter after {max_retries} retries.")
                 complete_filter = repair_query(complete_filter, str(e), input_json, openai_api_key)
-        pbar.close()
+        pbar_validation.close()
```
I see you changed the `pbar` variable but would be great if you ended the whitespace changes as well

areibman commented on requirements.txt:
```diff
@@ -1,3 +1,4 @@
-jsonschema==4.21.1
\ No newline at end of file
+jsonschema==4.21.1
+burr
```
Pin a version if possible

areibman commented on jaiqu/jaiqu_burr_straight_port.py:
This might make more sense as a Jupyter notebook example. Would prefer that we don't copy + paste the main code form Jaiqu.py 3 separate times for slightly different implementations.

areibman commented on jaiqu/jaiqu_burr_granular.py:
This might make more sense as a Jupyter notebook example. Would prefer that we don't copy + paste the main code form Jaiqu.py 3 separate times for slightly different implementations.

areibman commented on jaiqu/jaiqu_burr_granular.py:
```diff
@@ -0,0 +1,237 @@
+
+
+if __name__ == '__main__':
+    # app = build_application()
+    # app.visualize(
+    #     output_file_path="jaiqu_granular", include_conditions=True, view=True, format="png"
+    # )
```
Stated above, but main code with dummy data is probably better served as an example file.

skrawcz commented:
> @skrawcz seems you've made a lot of changes. Can you clarify/amend:
> 
> 1. Why are there so many .xml files? Are they necessary for this?
> 2. Could you either add a lint rc into the repo or undo all the changes where whitespace is getting changed but not affecting the code?
> 3. Update gitignore for files that don't need to be in there
> 4. I'd prefer a direct but optional implementation on top of Jaiqu.py if you're going to add Burr. It doesn't seem like a good idea to copy + paste 1 file into 3 separate ones. That, or create an examples directory for each.
> 5. Which implementation is supposed to be better? Granular or straight?
> 6. Would prefer any .PNGs just be rendered in README rather than just left in repo
> 
> Looks like a great idea, and it'd do a lot to help allay the complexity of Jaiqu.

Thanks for reviewing! This was a very quick and dirty PR -- as judged by the extra cruft that made it in. 😅 .

Part of this PR was to get your thoughts/feedback on it. I left the original jaiqu implementation in there for context as I was developing -- an actual implementation would as you suggest not duplicate all that code over and over. In terms of implementation is there one you prefer more? I was more or less mindlessly porting rather than thinking about an ideal structure...

skrawcz committed:
Makes Jaiqu use Burr

[Burr](https://github.com/dagworks-inc/burr) will enable someone to more easily debug
the state of the jaiqu application if it's giving nonsense back
or it hits the maxium number of retries.

The design decision here is to make a close port of the original
code. It could be broken down further, but that would require
more work and I wasn't sure it TQDM was something that
had to stay or not.

This is a 100% backwards compatible change with the
current API.

tests have been moved to live under /tests/ and match
the README.

The mermaid image has been replaced with the
actual application structure that Burr generates.

skrawcz committed:
Merge remote-tracking branch 'upstream/main'

# Conflicts:
#	jaiqu/jaiqu.py
#	pyproject.toml
#	requirements.txt

skrawcz committed:
Adds save_trace option to CLI

So that the CLI has parity with the Burr changes.

skrawcz commented:
> @skrawcz seems you've made a lot of changes. Can you clarify/amend:
> 
> 1. Why are there so many .xml files? Are they necessary for this?
> 2. Could you either add a lint rc into the repo or undo all the changes where whitespace is getting changed but not affecting the code?
> 3. Update gitignore for files that don't need to be in there
> 4. I'd prefer a direct but optional implementation on top of Jaiqu.py if you're going to add Burr. It doesn't seem like a good idea to copy + paste 1 file into 3 separate ones. That, or create an examples directory for each.
> 5. Which implementation is supposed to be better? Granular or straight?
> 6. Would prefer any .PNGs just be rendered in README rather than just left in repo
> 
> Looks like a great idea, and it'd do a lot to help allay the complexity of Jaiqu.

@areibman I think I fixed things up. As far as I have tested things work equivalently! 

I would squash merge this change and have a commit like:

```
[Burr](https://github.com/dagworks-inc/burr) will enable someone to more easily debug
the state of the jaiqu application if it's giving nonsense back
or it hits the maxium number of retries.

The design decision here is to make a close port of the original
code. It could be broken down further, but that would require
more work and I wasn't sure it TQDM was something that
had to stay or not.

This is a 100% backwards compatible change with the
current API.

tests have been moved to live under /tests/ and match
the README.

The mermaid image has been replaced with the
actual application structure that Burr generates.
```

skrawcz committed:
Fixes up typing and docs on tranlate_schema

Missed this in the merge.
